src/native/containers/dn-umap.h

```diff
@@ -0,0 +1,319 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#ifndef __DN_UMAP_H__
+#define __DN_UMAP_H__
+
+#include "dn-utils.h"
+#include "dn-allocator.h"
+
+typedef uint32_t (DN_CALLBACK_CALLTYPE *dn_umap_hash_func_t) (const void *key);
+typedef bool (DN_CALLBACK_CALLTYPE *dn_umap_equal_func_t) (const void *a, const void *b);
+typedef void (DN_CALLBACK_CALLTYPE *dn_umap_key_dispose_func_t) (void *key);
+typedef void (DN_CALLBACK_CALLTYPE *dn_umap_value_dispose_func_t) (void *value);
+typedef void (DN_CALLBACK_CALLTYPE *dn_umap_key_value_func_t) (void *key, void *value, void *user_data);
+
+typedef struct _dn_umap_node_t dn_umap_node_t;
+struct _dn_umap_node_t {
+	void *key;
+	void *value;
+	dn_umap_node_t *next;
+};
+
+typedef struct _dn_umap_t dn_umap_t;
+struct _dn_umap_t {
+	struct {
+		dn_umap_node_t **_buckets;
+		dn_umap_hash_func_t _hash_func;
+		dn_umap_equal_func_t _key_equal_func;
+		dn_umap_key_dispose_func_t _key_dispose_func;
+		dn_umap_value_dispose_func_t _value_dispose_func;
+		dn_allocator_t *_allocator;
+		uint32_t _bucket_count;
+		uint32_t _node_count;
+		uint32_t _threshold;
+		uint32_t _last_rehash;
+	} _internal;
+};
+
+typedef struct _dn_umap_it_t dn_umap_it_t;
+struct _dn_umap_it_t
+{
+	struct {
+		dn_umap_t *_map;
+		dn_umap_node_t *_node;
+		uint32_t _index;
+	} _internal;
+};
+
+typedef struct _dn_umap_result_t dn_umap_result_t;
+struct _dn_umap_result_t {
+	dn_umap_it_t it;
+	bool result;
+};
+
+typedef struct _dn_umap_custom_params_t dn_umap_custom_alloc_params_t;
+typedef struct _dn_umap_custom_params_t dn_umap_custom_init_params_t;
+struct _dn_umap_custom_params_t {
+	dn_allocator_t *allocator;
+	dn_umap_hash_func_t hash_func;
+	dn_umap_equal_func_t equal_func;
+	dn_umap_key_dispose_func_t key_dispose_func;
+	dn_umap_value_dispose_func_t value_dispose_func;
+};
+
+dn_umap_t *
+_dn_umap_alloc (
+	dn_allocator_t *allocator,
+	dn_umap_hash_func_t hash_func,
+	dn_umap_equal_func_t equal_func,
+	dn_umap_key_dispose_func_t key_dispose_func,
+	dn_umap_value_dispose_func_t value_dispose_func);
```

Comment:
Why do we need this?  Isn't `dn_unmap_alloc_custom` usable as the "low-level" interface that could be used by `dn_unmap_alloc`?

Reply:
yes, we could get rid of that variation.